File: tmdb-core/src/main/java/dev/reuss/tmdb/TmdbClientBuilder.java
package dev.reuss.tmdb;

import dev.reuss.tmdb.core.auth.TmdbAuth;
import dev.reuss.tmdb.core.config.TmdbClientConfig;
import dev.reuss.tmdb.core.metrics.TmdbMetricsRecorder;
import dev.reuss.tmdb.value.language.Language;
import dev.reuss.tmdb.value.language.Languages;
import dev.reuss.tmdb.value.region.Region;

import java.time.Duration;
import java.util.Objects;

/**
 * Builder for creating {@link TmdbClient} instances.
 *
 * <p>The builder configures authentication, the TMDB API base URL and default
 * request settings such as language, region and timeouts.</p>
 *
 * <p>At minimum, an access token or {@link TmdbAuth} instance must be provided
 * before calling {@link #build()}.</p>
 *
 * <pre>{@code
 * TmdbClient tmdb = TmdbClient.builder()
 *         .accessToken("your-access-token")
 *         .defaultLanguage(Languages.DE_DE)
 *         .build();
 * }</pre>
 *
 * @see TmdbClient
 * @see TmdbAuth
 * @see TmdbClientConfig
 */
public final class TmdbClientBuilder {

    private TmdbAuth auth;
    private String baseUrl = TmdbClientConfig.DEFAULT_BASE_URL;
    private Language defaultLanguage = Languages.EN_US;
    private Region defaultRegion;
    private Duration connectTimeout = TmdbClientConfig.DEFAULT_CONNECT_TIMEOUT_DURATION;
    private Duration requestTimeout = TmdbClientConfig.DEFAULT_REQUEST_TIMEOUT_DURATION;
    private TmdbMetricsRecorder metricsRecorder = TmdbMetricsRecorder.NOOP;

    /**
     * Sets the TMDB bearer access token.
     *
     * @param accessToken the TMDB bearer access token
     * @return this builder
     * @throws IllegalArgumentException if the access token is {@code null}, blank or empty
     */
    public TmdbClientBuilder accessToken(String accessToken) {
        this.auth = TmdbAuth.bearerToken(accessToken);
        return this;
    }

    /**
     * Sets the TMDB authentication configuration.
     *
     * @param auth the TMDB authentication configuration
     * @return this builder
     * @throws NullPointerException if {@code auth} is {@code null}
     */
    public TmdbClientBuilder auth(TmdbAuth auth) {
        this.auth = Objects.requireNonNull(auth, "TMDB auth must not be null");
        return this;
    }

    /**
     * Sets the TMDB API base URL.
     *
     * <p>The default value is {@code https://api.themoviedb.org/3}.
     * This is mainly useful for testing or custom deployments.</p>
     *
     * @param baseUrl the TMDB API base URL
     * @return this builder
     */
    public TmdbClientBuilder baseUrl(String baseUrl) {
        this.baseUrl = baseUrl;
        return this;
    }

    /**
     * Sets the default language for localized TMDB requests.
     *
     * <p>The default value is {@link Languages#EN_US}.</p>
     *
     * @param defaultLanguage the default language
     * @return this builder
     */
    public TmdbClientBuilder defaultLanguage(Language defaultLanguage) {
        this.defaultLanguage = defaultLanguage;
        return this;
    }

    /**
     * Sets the default region for regional TMDB requests.
     *
     * <p>The default region is optional and may be {@code null}. Not every
     * TMDB request requires or supports a region.</p>
     *
     * @param defaultRegion the default region, or {@code null}
     * @return this builder
     */
    public TmdbClientBuilder defaultRegion(Region defaultRegion) {
        this.defaultRegion = defaultRegion;
        return this;
    }

    /**
     * Sets the HTTP connection timeout.
     *
     * <p>The default value is 5 seconds.</p>
     *
     * @param connectTimeout the connection timeout
     * @return this builder
     */
    public TmdbClientBuilder connectTimeout(Duration connectTimeout) {
        this.connectTimeout = connectTimeout;
        return this;
    }

    /**
     * Sets the overall HTTP request timeout.
     *
     * <p>The default value is 10 seconds.</p>
     *
     * @param requestTimeout the request timeout
     * @return this builder
     */
    public TmdbClientBuilder requestTimeout(Duration requestTimeout) {
        this.requestTimeout = requestTimeout;
        return this;
    }

    /**
     * Sets the metrics recorder used to observe TMDB HTTP requests.
     *
     * <p>The default recorder is a no-op implementation.</p>
     *
     * @param metricsRecorder the metrics recorder
     * @return this builder
     * @throws NullPointerException if {@code metricsRecorder} is {@code null}
     */
    public TmdbClientBuilder metricsRecorder(TmdbMetricsRecorder metricsRecorder) {
        this.metricsRecorder = Objects.requireNonNull(metricsRecorder, "metricsRecorder must not be null");
        return this;
    }

    /**
     * Builds a new {@link TmdbClient}.
     *
     * @return a new TMDB client instance
     * @throws NullPointerException     if required configuration values are missing
     * @throws IllegalArgumentException if configuration values are invalid
     */
    public TmdbClient build() {
        TmdbClientConfig config = new TmdbClientConfig(
                auth,
                baseUrl,
                defaultLanguage,
                defaultRegion,
                connectTimeout,
                requestTimeout,
                metricsRecorder
        );

        return new DefaultTmdbClient(config);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/TmdbClientBuilder.kt
package dev.reuss.tmdb

import dev.reuss.tmdb.core.auth.TmdbAuth
import dev.reuss.tmdb.core.config.TmdbClientConfig
import dev.reuss.tmdb.core.metrics.TmdbMetricsRecorder
import dev.reuss.tmdb.value.language.Language
import dev.reuss.tmdb.value.language.Languages
import dev.reuss.tmdb.value.region.Region
import java.time.Duration

/**
 * Configures the default [TmdbClient] implementation.
 *
 * The builder configures authentication, the TMDB API base URL and default
 * request settings such as language, region and timeouts.
 *
 * A bearer access token or [TmdbAuth] is required. Unless overridden, requests
 * use TMDB API v3, `en-US`, a five-second connection timeout, a ten-second
 * request timeout, no default region, and no-op metrics.
 *
 * Example:
 *
 * ```
 * val tmdb = TmdbClient.builder()
 *     .accessToken("your-access-token")
 *     .defaultLanguage(Languages.DE_DE)
 *     .build()
 * ```
 *
 * @see TmdbClient
 * @see TmdbAuth
 * @see TmdbClientConfig
 */
class TmdbClientBuilder {
    private var auth: TmdbAuth? = null
    private var baseUrl: String = TmdbClientConfig.DEFAULT_BASE_URL
    private var defaultLanguage: Language = Languages.EN_US
    private var defaultRegion: Region? = null
    private var connectTimeout: Duration = TmdbClientConfig.DEFAULT_CONNECT_TIMEOUT_DURATION
    private var requestTimeout: Duration = TmdbClientConfig.DEFAULT_REQUEST_TIMEOUT_DURATION
    private var metricsRecorder: TmdbMetricsRecorder = TmdbMetricsRecorder.NOOP

    /**
     * Uses [accessToken] for bearer authentication.
     *
     * The token is trimmed and must not be blank.
     *
     * @throws IllegalArgumentException if [accessToken] is blank
     */
    fun accessToken(accessToken: String): TmdbClientBuilder =
        apply {
            auth = TmdbAuth.bearerToken(accessToken)
        }

    /** Uses an already validated authentication value. */
    fun auth(auth: TmdbAuth): TmdbClientBuilder =
        apply {
            this.auth = auth
        }

    /**
     * Overrides the TMDB API base URL.
     *
     * The default value is `https://api.themoviedb.org/3`.
     * The value is validated by [build] as an absolute HTTP(S) URL without a
     * query or fragment. This setting is mainly useful for tests and compatible
     * proxy deployments.
     */
    fun baseUrl(baseUrl: String): TmdbClientBuilder =
        apply {
            this.baseUrl = baseUrl
        }

    /**
     * Sets the language added to requests that do not override `language`.
     *
     * The default value is [Languages.EN_US].
     */
    fun defaultLanguage(defaultLanguage: Language): TmdbClientBuilder =
        apply {
            this.defaultLanguage = defaultLanguage
        }

    /**
     * Sets the region added to requests that do not override `region`.
     *
     * The default is `null`, which omits the parameter.
     */
    fun defaultRegion(defaultRegion: Region?): TmdbClientBuilder =
        apply {
            this.defaultRegion = defaultRegion
        }

    /**
     * Sets the maximum time allowed to establish an HTTP connection.
     *
     * The default is five seconds. [build] rejects zero and negative durations.
     */
    fun connectTimeout(connectTimeout: Duration): TmdbClientBuilder =
        apply {
            this.connectTimeout = connectTimeout
        }

    /**
     * Sets the timeout applied to each HTTP request.
     *
     * The default is ten seconds. [build] rejects zero and negative durations.
     */
    fun requestTimeout(requestTimeout: Duration): TmdbClientBuilder =
        apply {
            this.requestTimeout = requestTimeout
        }

    /**
     * Sets the observer notified about TMDB HTTP request lifecycle events.
     *
     * The default is [TmdbMetricsRecorder.NOOP].
     */
    fun metricsRecorder(metricsRecorder: TmdbMetricsRecorder): TmdbClientBuilder =
        apply {
            this.metricsRecorder = metricsRecorder
        }

    /**
     * Validates the accumulated configuration and creates a client.
     *
     * @throws IllegalArgumentException if authentication is missing or the base URL
     * or either timeout is invalid
     */
    fun build(): TmdbClient {
        val auth =
            requireNotNull(auth) {
                "TMDB auth must not be null"
            }

        val config =
            TmdbClientConfig(
                auth,
                baseUrl,
                defaultLanguage,
                defaultRegion,
                connectTimeout,
                requestTimeout,
                metricsRecorder,
            )

        return DefaultTmdbClient(config)
    }
}
